### PR TITLE
Improve performance and memory for validation by caching XML Schema / DTD

### LINKED ISSUE #534: Improve performance and memory for validation by caching XML Schema / DTD
Today each time completion / validation based on XML Schema / DTD, the grammar is parsed.

We should cache those grammar to use it in completion and validation both (if it's possible) and reset the cache when file of the grammar is updated.

### PR DESCRIPTION
**Pull request by angelozerr**
Fixes #534

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
It works with simple case but all tests are failed. It's definitely more hard than I though. I must work on again on this issue...

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/FileChangedTracker.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package org.eclipse.lsp4xml.extensions.contentmodel.model;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/FileChangedTracker.java]: copyright header

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java [hunk 2]
```diff
@@ -0,0 +1,419 @@
+/*
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java]: where is this file coming from? need to document it

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+	 * 
+	 * @param grammars An array containing the set of grammars being returned; order
+	 * is not significant.
+	 */
+	@Override
+	public void cacheGrammars(String grammarType, Grammar[] grammars) {
+		if (!fPoolIsLocked) {
+			for (int i = 0; i < grammars.length; i++) {
+				if (DEBUG) {
+					System.out.println("CACHED GRAMMAR " + (i + 1));
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java]: use logger

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+					Grammar temp = grammars[i];
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java]: unused variable

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+	 *
+	 * @param desc The description of the Grammar being requested.
+	 * 
+	 * @return The Grammar corresponding to this description or null if no such
+	 * Grammar is known.
+	 */
+	@Override
+	public Grammar retrieveGrammar(XMLGrammarDescription desc) {
+		if (DEBUG) {
+			System.out.println("RETRIEVING GRAMMAR FROM THE APPLICATION WITH FOLLOWING DESCRIPTION :");
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java]: use logger

**Comment by angelozerr** [comment]
After fighting with Xerces, I think (I hope) I have a cleaner implementation for validation cache. 

@xorye @fbricon please play with this PR. We have a lot of tests but please test those following usecases:

 * XML file bind with XML Schema with schemaLocation. Update the XML Schema and check validation continue to work.
 * XML file bind with XML Schema A which includes a XMl Schema B with schemaLocation. Update the XML Schema B and check validation continue to work.

 * Do the same tests with XML file bound with XML Schema with noSchemaLocationNamespace.
 * Open a second XML file which bound an another XML Schema  with noSchemaLocationNamespace. and tries if it works (It was a Xerces limitation that I figure it, I must write tests for that).
 * Do the same tests with XML file bound with DTD

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/utils/XSDUtils.java [hunk 6]
```diff
@@ -345,4 +351,54 @@ public static boolean isXSTargetElement(Element element) {
 	public static boolean isXSAttribute(DOMElement element) {
 		return "attribute".equals(element.getLocalName());
 	}
+
+	public static FilesChangedTracker createFilesChangedTracker(SchemaGrammar grammar) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/utils/XSDUtils.java]: return createFilesChangedTracker(Collections.singleton(grammar));

File: org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLValidationPoolCacheTest.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+		xml = "<Annotation\r\n" + "	xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\r\n"
+				+ "	xsi:noNamespaceSchemaLocation=\"src/test/resources/xsd/pattern.xsd\"\r\n"
+				+ "	Term=\"X\"></Annotation>";
+		Diagnostic patternValid = d(3, 6, 3, 9, XMLSchemaErrorCode.cvc_pattern_valid);
+		Diagnostic cvcAttribute3 = d(3, 6, 3, 9, XMLSchemaErrorCode.cvc_attribute_3);
+		testDiagnosticsFor(xmlLanguageService, xml, patternValid, cvcAttribute3);
+	}
+	
+	@Test
+	public void schemaLocation() {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/java/org/eclipse/lsp4xml/extensions/contentmodel/XMLValidationPoolCacheTest.java]: useless method

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/FileChangedTracker.java (hunk 1 above)
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/FileChangedTracker.java]: This class was removed.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java (hunk 2 above)
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java]: fixed

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java (hunk 4 above)
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java]: This class has been cleaned.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java (hunk 5 above)
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/diagnostics/LSPXMLGrammarPool.java]: This class has been cleaned.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/utils/XSDUtils.java (hunk 6 above)
Comment (angelozerr) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/utils/XSDUtils.java]: fixed

**Pull request merged**